### PR TITLE
planner: use right hint to build join order | tidb-test=pr/2463

### LINKED ISSUE #58325: Choose left side as build in semi and anti semi join in optimizer
## Enhancement

Currently, hash join can only choose right side as build table in semi and anti-semi join. For example, for the sql `select * from t1 where exists (select 1 from t2 where t1.col0 = t2.col0) order by t1.col0, t1.col1;`, if t2 has 1000w rows, t1 has 1000 rows. It's obvious that choose t1 as build table is better, but we don't support to build table for left side in previous hash join implementation. After [#57658](https://github.com/pingcap/tidb/pull/57658) and [#57971](https://github.com/pingcap/tidb/pull/57971) merge, hash join v2 could support left side as build table.  So we need to make optimizer enable to choose left side as build table when necessary.

### LINKED ISSUE #58571: Hash join statistic is incorrect when spill is triggered
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Trigger spill in hash join.

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

correct statistic

### 3. What did you see instead (Required)

```
+------------------------------------+--------------+-----------+-----------+----------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+----------+---------+
| id                                 | estRows      | actRows   | task      | access object  | execution info | operator info | memory   | disk    |
+------------------------------------+--------------+-----------+-----------+----------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+----------+---------+
| Sort_10                            | 1.00         | 5         | root      |                | time:1m15.8s, open:657.4µs, close:29.5µs, loops:2, RU:1199689.70                                                                                                                                                                                                                                                                         | test.orders.o_orderpriority                                                                                                               | 1.21 KB  | 0 Bytes |
| └─Projection_12                    | 1.00         | 5         | root      |                | time:1m15.8s, open:645.4µs, close:28.6µs, loops:5, Concurrency:OFF | test.orders.o_orderpriority, Column#26                                                                                                    | 34.0 KB  | N/A     |
|   └─HashAgg_13                     | 1.00         | 5         | root      |                | time:1m15.8s, open:641.8µs, close:27µs, loops:5, partial_worker:{wall_time:1m15.813126507s, concurrency:5, task_num:2588, tot_wait:6m18.325465785s, tot_exec:611.285923ms, tot_time:6m19.065554441s, max:1m15.813112828s, p95:1m15.813112828s}, final_worker:{wall_time:1m15.813146976s, concurrency:5, task_num:21, tot_wait:16.761µs, tot_exec:13.302µs, tot_time:6m19.065664816s, max:1m15.813135023s, p95:1m15.813135023s}                                             | group by:test.orders.o_orderpriority, funcs:count(1)->Column#26, funcs:firstrow(test.orders.o_orderpriority)->test.orders.o_orderpriority | 481.9 KB | 0 Bytes |
|     └─HashJoin_26                  | 2281928.20   | 2574425   | root      |                | time:1m15.8s, open:585.6µs, close:24.4µs, loops:2589, build_hash_table:{total:160.9ms, fetch:-7.013241164s, build:7.51s}, probe:{concurrency:5, total:3m25s, max:40.5s, probe:3.09s, fetch and wait:3m21.9s, probe_collision:19336081}                                                                                                                                                                                                                                     | semi join, left side:TableReader_29, equal:[eq(test.orders.o_orderkey, test.lineitem.l_orderkey)]                                         | 1.00 GB  | 9.45 GB |                                                                                                                                                       
|       ├─TableReader_32(Build)      | 240004648.80 | 189670348 | root      |                | time:17.3s, open:425.7µs, close:3.61µs, loops:184597, cop_task: {num: 5628, max: 309.5ms, min: 268.9µs, avg: 58.1ms, p95: 143.3ms, max_proc_keys: 80864, p95_proc_keys: 79840, tot_proc: 5m15.4s, tot_wait: 747.8ms, copr_cache_hit_ratio: 0.00, build_task_duration: 383.2µs, max_distsql_concurrency: 15}, rpc_info:{Cop:{num_rpc:5628, total_time:5m26.7s}}                                                                                                             | data:Selection_31 | 18.5 MB  | N/A     |                                                                                                                                                       
|       │ └─Selection_31             | 240004648.80 | 189670348 | cop[tikv] |                | tikv_task:{proc max:0s, min:0s, avg: 54.1ms, p80:78ms, p95:133ms, iters:315458, tasks:5628}, scan_detail: {total_process_keys: 299982707, total_process_keys_size: 58990940562, total_keys: 299988321, get_snapshot_time: 403.4ms, rocksdb: {key_skipped_count: 299982707, block: {cache_hit_count: 1883561}}}, time_detail: {total_process_time: 5m15.4s, total_suspend_time: 1.11s, total_wait_time: 747.8ms, total_kv_read_wall_time: 4m37.7s, tikv_wall_time: 5m18.8s} | lt(test.lineitem.l_commitdate, test.lineitem.l_receiptdate)                                                                               | N/A      | N/A     |                                                                                                                                                       
|       │   └─TableFullScan_30       | 300005811.00 | 300005811 | cop[tikv] | table:lineitem | tikv_task:{proc max:0s, min:0s, avg: 49.4ms, p80:71.6ms, p95:121.7ms, iters:315458, tasks:5628}                                                                                                                                                                                                                                                                                                                                                                            | keep order:false | N/A      | N/A     |
|       └─TableReader_29(Probe)      | 2852410.25   | 2807525   | root      |                | time:4.4s, open:143.5µs, close:15.2µs, loops:2734, cop_task: {num: 395, max: 1.04s, min: 214.8µs, avg: 150.6ms, p95: 658.5ms, max_proc_keys: 879577, p95_proc_keys: 871392, tot_proc: 59s, tot_wait: 54.9ms, copr_cache_hit_ratio: 0.21, build_task_duration: 81.4µs, max_distsql_concurrency: 15}, rpc_info:{Cop:{num_rpc:395, total_time:59.5s}}                                                                                                                         | data:Selection_28 | 2.08 MB  | N/A     |                                                                                                                                                       
|         └─Selection_28             | 2852410.25   | 2807525   | cop[tikv] |                | tikv_task:{proc max:1.02s, min:1ms, avg: 151.1ms, p80:291ms, p95:652ms, iters:74858, tasks:395}, scan_detail: {total_process_keys: 74173248, total_process_keys_size: 11266342171, total_keys: 74173559, get_snapshot_time: 39.4ms, rocksdb: {key_skipped_count: 74173248, block: {cache_hit_count: 376619}}}, time_detail: {total_process_time: 59s, total_suspend_time: 115.9ms, total_wait_time: 54.9ms, total_kv_read_wall_time: 54.6s, tikv_wall_time: 59.2s}         | ge(test.orders.o_orderdate, 1995-01-01 00:00:00.000000), lt(test.orders.o_orderdate, 1995-04-01 00:00:00.000000)                          | N/A      | N/A     |                                                                                                                                                       
|           └─TableFullScan_27       | 75000000.00  | 75000000  | cop[tikv] | table:orders   | tikv_task:{proc max:970ms, min:1ms, avg: 141.5ms, p80:275ms, p95:611ms, iters:74858, tasks:395} | keep order:false                                                                                                                          | N/A      | N/A     |                                                                                                                                                       
+------------------------------------+--------------+-----------+-----------+----------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------+----------+---------+
10 rows in set (1 min 15.82 sec)
```

### 4. What is your TiDB version? (Required)

nightly(9b70321)

<!-- Paste the output of SELECT tidb_version() -->

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: close #58325

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: tests/integrationtest/r/tpch.result [hunk 1]
```diff
@@ -1294,8 +1294,8 @@ Sort	1.00	root		Column#31
...
-        ├─TableReader(Build)	75000000.00	root		data:TableFullScan
-        │ └─TableFullScan	75000000.00	cop[tikv]	table:orders	keep order:false
-        └─TableReader(Probe)	0.80	root		data:Selection
-          └─Selection	0.80	cop[tikv]		gt(tpch50.customer.c_acctbal, NULL), in(substring(tpch50.customer.c_phone, 1, 2), "20", "40", "22", "30", "39", "42", "21")
-            └─TableFullScan	7500000.00	cop[tikv]	table:customer	keep order:false
+        ├─TableReader(Build)	0.80	root		data:Selection
+        │ └─Selection	0.80	cop[tikv]		gt(tpch50.customer.c_acctbal, NULL), in(substring(tpch50.customer.c_phone, 1, 2), "20", "40", "22", "30", "39", "42", "21")
+        │   └─TableFullScan	7500000.00	cop[tikv]	table:customer	keep order:false
+        └─TableReader(Probe)	75000000.00	root		data:TableFullScan
+          └─TableFullScan	75000000.00	cop[tikv]	table:orders	keep order:false
```
Comment (hawkingrei) [inline tests/integrationtest/r/tpch.result:1301]: it shows the bug. the origin is wrong.

Comment (hawkingrei) [reply tests/integrationtest/r/tpch.result:1301]: although it is the bug, ```TestSemiJoinOrder``` is right.

**Review by windtalker — APPROVED** [review]
lgtm

**Review by AilinKid — APPROVED** [review]

**Pull request merged**